### PR TITLE
Add single-user authentication

### PR DESCRIPTION
**Pull request by O-Bezimo**
| Q               | A
| --------------- | ---
| Bug fix?             | no
| New feature?    | yes
| API breaks?       | no
| Deprecations?   | no
| Related tickets  | fixes #X, partially #Y, mentioned in #Z
| License         | Apache 2.0

### What's in this PR?

Our objective is to add the Single-User authentication in addition to the OIDC and LDAP authentication methods.

We added a SingleUserConfiguration object in the NifiCluster ressource.
The options are the following:
- Enabled boolean: set to true to add the according single-user configuration blocs in the nifi.properties template and the login_identity_provider template.
- SecretRef SecretReference: reference to a kubernetes secret containing the desired username and password.
- AuthorizerEnabled bool: set to true to use the single-user-authorizer. Set to false to use the default managed-authorizer.

We also added the reference to the SingleUserConfiguration in the helm folder.

### Why?

If neither OIDC or LDAP are configured, nifi forces a certificate authentication on all HTTP/HTTPS certificates.
We need another option so that we can have functionnal healthchecks on our loadbalancer generated by the ingress-controller.

### Additional context

We tested the code with an integration cluster hosted on gcp and telepresence as a proxy for our modified operator.

We also proofed some error cases:
- if SingleUserConfiguration.Enabled is false and SingleUserConfiguration.AuthorizerEnabled is true: don't configure the authorizer
- if SecretRef is empty: use auto generated username and password that would appear in the logs

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](docs/error-handling-guide.md)
- [x] Logging code meets the guideline
- [x] User guide and development docs updated (if needed)
- [x] Append changelog with changes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -10,11 +10,17 @@
 
 ### Removed
 
+## v1.3.0
+
+### Added
+
+- [PR #278](https://github.com/konpyutaika/nifikop/pull/278) - **[Operator]** Added the single-user-authentication method.
```
Comment (juldrixx) [inline CHANGELOG.md:17]:
```suggestion
## Unreleased

### Added

- [PR #278](https://github.com/konpyutaika/nifikop/pull/278) - **[Operator]** Added the single-user-authentication method.

### Changed

### Fixed Bugs

### Deprecated

### Removed
```

The next release number is not defined right now.

**Review by juldrixx — COMMENTED** [review]
Please delete the versionned documentation, the next version has not yet been decided.

And add the appropriate changes to the "unversioned" documentation to reflect your changes.

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,11 +10,17 @@
...
+
```
Comment (juldrixx) [inline CHANGELOG.md:18]:
```suggestion
```suggestion
## Unreleased

### Added
- [PR #278](https://github.com/konpyutaika/nifikop/pull/278) - **[Operator]** Added the single-user-authentication method.

### Changed

### Fixed Bugs

### Deprecated

### Removed

```
```

**Comment by juldrixx** [comment]
@mh013370 Should secret keys be supplied in the operator configuration or forced as now?

File: CHANGELOG.md [hunk 3]
```diff
@@ -10,11 +10,17 @@
...
 ## v1.2.0
 
 ### Added
 
-- [PR #258](https://github.com/konpyutaika/nifikop/pull/141) - **[Helm Chart]** Upgraded helm-deployed HPA to v2 and added flowPosition to NiFiDataflow.
+- [PR #258](https://github.com/konpyutaika/nifikop/pull/141) - **[Operator]** Upgraded helm-deployed HPA to v2 and added flowPosition to NiFiDataflow.
```
[the remark this answers is not in the document]
Comment (O-Bezimo) [reply CHANGELOG.md]: Modified accordingly

File: CHANGELOG.md (hunk 2 above)
Comment (O-Bezimo) [reply CHANGELOG.md]: Modified accordingly

**Comment by O-Bezimo** [comment]
> @mh013370 Should secret keys be supplied in the operator configuration or forced as now?

Secret keys are supplied if singleUserConfiguration->enabled is true and if singleUserConfiguration->secretRef is not given in the nificluster.yml configuration. If enabled is true and secretRef is declared and exists in the cluster with a username and password, the operator will use these instead as credentials and add it in the login_identity_providers file.

**Comment by mh013370** [comment]
> @mh013370 Should secret keys be supplied in the operator configuration or forced as now?



> > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> 
> Secret keys are supplied if singleUserConfiguration->enabled is true and if singleUserConfiguration->secretRef is not given in the nificluster.yml configuration. If enabled is true and secretRef is declared and exists in the cluster with a username and password, the operator will use these instead as credentials and add it in the login_identity_providers file.

If `SingleUserConfiguration.SecretRef` is optional, then it should be expressed as such in the spec by using a pointer (allowing you to check for `nil`) and adding the `// +optional` kubebuilder annotation.

**Comment by mh013370** [comment]
Why are there a whole new set of `unversioned` docs?

File: api/v1/nificluster_types.go [hunk 4]
```diff
@@ -100,11 +100,18 @@ type NifiClusterSpec struct {
...
 	// ControllerUserIdentity specifies what to call the static admin user's identity
 	// Warning: once defined don't change this value either the operator will no longer be able to manage the cluster
-	ControllerUserIdentity *string `json:"controllerUserIdentity,omitempty"`
+	ControllerUserIdentity  *string                 `json:"controllerUserIdentity,omitempty"`
+	SingleUserConfiguration SingleUserConfiguration `json:"singleUserConfiguration,omitempty"`
 
 	// @TODO: Block Controller change
 }
 
+type SingleUserConfiguration struct {
```
Comment (mh013370) [inline api/v1/nificluster_types.go:109]: These all need comments briefly explaining what they are and what they're used for. Especially the `SecretRef` & what its contents should look like.

File: api/v1/nificluster_types.go [hunk 5]
```diff
@@ -100,11 +100,18 @@ type NifiClusterSpec struct {
...
+	Enabled           bool            `json:"enabled,omitempty"`
+	AuthorizerEnabled bool            `json:"authorizerEnabled,omitempty"`
+	SecretRef         SecretReference `json:"secretRef,omitempty"`
```
Comment (mh013370) [inline api/v1/nificluster_types.go:112]: Like mentioned in my other comment, recommend making this a pointer to a `SecretReference` and adding the `// +optional` kubebuilder annotation.

**Comment by O-Bezimo** [comment]
> > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> 
> > > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> > 
> > 
> > Secret keys are supplied if singleUserConfiguration->enabled is true and if singleUserConfiguration->secretRef is not given in the nificluster.yml configuration. If enabled is true and secretRef is declared and exists in the cluster with a username and password, the operator will use these instead as credentials and add it in the login_identity_providers file.
> 
> If `SingleUserConfiguration.SecretRef` is optional, then it should be expressed as such in the spec by using a pointer (allowing you to check for `nil`) and adding the `// +optional` kubebuilder annotation.

Thank you for your answer, the modification have been made

**Comment by O-Bezimo** [comment]
> > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> 
> > > @mh013370 Should secret keys be supplied in the operator configuration or forced as now?
> > 
> > 
> > Secret keys are supplied if singleUserConfiguration->enabled is true and if singleUserConfiguration->secretRef is not given in the nificluster.yml configuration. If enabled is true and secretRef is declared and exists in the cluster with a username and password, the operator will use these instead as credentials and add it in the login_identity_providers file.
> 
> If `SingleUserConfiguration.SecretRef` is optional, then it should be expressed as such in the spec by using a pointer (allowing you to check for `nil`) and adding the `// +optional` kubebuilder annotation.

A pointer is added to Secret ref and the empty checks are made with nil as of now. An optional annotation is added and the same is made to a secretKeys struct allowing the final user to change the username and password keys. They will default to "username" and "password" if nil and the according information are added in the documentation and in the helm folder.

**Comment by O-Bezimo** [comment]
> Why are there a whole new set of `unversioned` docs?

I switched the 1.3 versioned docs to unversioned which was a mistake. Since then I deleted it and moved the modifications in ./site/docs.

File: api/v1/nificluster_types.go (hunk 5 above)
Comment (O-Bezimo) [reply api/v1/nificluster_types.go]: Done

File: api/v1/nificluster_types.go (hunk 4 above)
Comment (O-Bezimo) [reply api/v1/nificluster_types.go:111]: Done

File: helm/nifi-cluster/values.yaml [hunk 6]
```diff
@@ -25,6 +25,19 @@ cluster:
...
+    enabled: false
+    authorizerEnabled: false
+    secretRef: {
+      "name": "single-user-credentials",
+      "namespace": "nifi"
+    }
+    secretKeys: {
+      username: "username",
+      password: "password"
+    }
```
Comment (juldrixx) [inline helm/nifi-cluster/values.yaml:39]:
```suggestion
    secretRef:
      name: "single-user-credentials"
      namespace: "nifi"
    secretKeys:
      username: "username"
      password: "password"
```

File: pkg/resources/templates/config/authorizers.go [hunk 7]
```diff
@@ -33,10 +41,19 @@ var AuthorizersTemplate = `{{- $nodeList := .NodeList }}
         <property name="Access Policy Provider">file-access-policy-provider</property>
     </authorizer>
 </authorizers>
+{{end}}
```
Comment (juldrixx) [inline pkg/resources/templates/config/authorizers.go:44]:
```suggestion
{{- end}}
```

File: pkg/resources/templates/config/authorizers.go [hunk 8]
```diff
@@ -56,6 +73,7 @@ var EmptyAuthorizersTemplate = `<authorizers>
...
```
Comment (juldrixx) [inline pkg/resources/templates/config/authorizers.go:76]:
```suggestion
{{- end}}
```

File: site/docs/5_references/1_nifi_cluster/1_nifi_cluster.md [hunk 9]
```diff
@@ -193,6 +194,15 @@ spec:
...
+| ----------------- | ---------------- | ----------------------------------------------------------------------------------------------------------------------------------------- | -------- | ------- |
+| enabled           | boolean          | if set to true, we will enable single-user usage into nifi.properties configuration.                                                                     | No       | false   |
+| authorizerEnabled | boolean          | if set to true, we will use the single-user-authorizer instead of the managed-authorizer.                                                                | No       | false   |
+| secretRef         | [*SecretReference](../4_nifi_parameter_context#secretreference)   | the reference to the username and password. If not configured, will use auto generated username and password that will appear once in the logs at startup.      | No       | nil     |
+| secretKeys         | *SecretKeys   | the keys of the username and password.      | No       | {username= "username", password="password"}     |
```
Comment (juldrixx) [inline site/docs/5_references/1_nifi_cluster/1_nifi_cluster.md:204]: A documentation section should be add for the SecretKeys doc to describe it. Like it's done for the SecretReference.
Directly in the 1_nifi_cluster.md file.

File: api/v1/nificluster_types.go [hunk 10]
```diff
@@ -101,10 +101,35 @@ type NifiClusterSpec struct {
...
+	// +optional
+	SecretRef *SecretReference `json:"secretRef,omitempty"`
+	// The keys referencing the username and password
+	// +optional
+	SecretKeys *SecretKeys `json:"secretKeys,omitempty"`
+}
+
+type SecretKeys struct {
+	// +kubebuilder:default:=username
+	Username string `json:"username,omitempty"`
```
[the remark this answers is not in the document]
Comment (O-Bezimo) [reply api/v1/nificluster_types.go:129]: Modified

File: api/v1/nificluster_types.go [hunk 11]
```diff
@@ -101,10 +101,35 @@ type NifiClusterSpec struct {
...
+	// +kubebuilder:default:=password
+	Password string `json:"password,omitempty"`
```
[the remark this answers is not in the document]
Comment (O-Bezimo) [reply api/v1/nificluster_types.go:132]: Modified

File: helm/nifi-cluster/templates/nifi-cluster.yaml [hunk 12]
```diff
@@ -22,6 +22,19 @@ spec:
...
   {{ end }}
+  singleUserConfiguration:
+    {{ if .Values.cluster.singleUserConfiguration.enabled }}
+      enabled: {{ .Values.cluster.enabled }}
+    {{ end }}
+    {{ if .Values.cluster.singleUserConfiguration.secretRef }}
+      secretRef: {{ toYaml .Values.cluster.singleUserConfiguration.secretRef | nindent 6 }}
+    {{ end }}
+    {{ if .Values.cluster.singleUserConfiguration.secretKeys }}
+      secretRef: {{ toYaml .Values.cluster.singleUserConfiguration.secretKeys | nindent 6 }}
```
[the remark this answers is not in the document]
Comment (O-Bezimo) [reply helm/nifi-cluster/templates/nifi-cluster.yaml]: Modified

File: helm/nifi-cluster/values.yaml (hunk 6 above)
Comment (O-Bezimo) [reply helm/nifi-cluster/values.yaml]: Modified

File: pkg/resources/templates/config/authorizers.go (hunk 7 above)
Comment (O-Bezimo) [reply pkg/resources/templates/config/authorizers.go]: Modified

File: pkg/resources/templates/config/authorizers.go (hunk 8 above)
Comment (O-Bezimo) [reply pkg/resources/templates/config/authorizers.go]: Modified

File: site/docs/5_references/1_nifi_cluster/1_nifi_cluster.md (hunk 9 above)
Comment (O-Bezimo) [reply site/docs/5_references/1_nifi_cluster/1_nifi_cluster.md]: Section added and link to the new section made in singleuserconfiguration

File: helm/nifi-cluster/values.yaml [hunk 13]
```diff
@@ -25,6 +25,19 @@ cluster:
...
+    enabled: false
+    authorizerEnabled: false
+    secretRef: {
+      name: "single-user-credentials",
+      namespace: "nifi"
+    }
+    secretKeys: {
+      username: "username",
+      password: "password"
+    }
```
Comment (juldrixx) [inline helm/nifi-cluster/values.yaml:39]:
```suggestion
    secretRef:
      name: "single-user-credentials"
      namespace: "nifi"
    secretKeys:
      username: "username"
      password: "password"
```

File: helm/nifi-cluster/values.yaml (hunk 6 above)
Comment (juldrixx) [reply helm/nifi-cluster/values.yaml]: It has not.

**Review by juldrixx — APPROVED** [review]

**Pull request merged**